{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7e81cbd9-1811-41a7-a310-23e6e321b00e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded AnnData object: AnnData object with n_obs × n_vars = 2004 × 34346\n",
      "    obs: 'batch', 'experiment', 'technology', 'ct', 'ct_orig', 'stage', 'timepoint', 'n_genes_by_counts', 'total_counts', 'total_counts_mt', 'pct_counts_mt', 'n_genes', 'n_counts'\n",
      "    var: 'gene_ids', 'gene_symbol', 'mt', 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts', 'n_cells'\n",
      "    uns: 'log1p'\n",
      "    layers: 'counts'\n",
      "['16C' '8C' '4C' '2C' 'Zygote' 'TE' 'ICM' 'EPI' 'PrE']\n"
     ]
    }
   ],
   "source": [
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.gridspec as gridspec\n",
    "import os\n",
    "from scipy.stats import ttest_ind\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "from statannotations.Annotator import Annotator\n",
    "import warnings\n",
    "warnings.simplefilter(\"ignore\", UserWarning)\n",
    "\n",
    "# Check if adata is defined\n",
    "if 'adata' not in locals() and 'adata' not in globals():\n",
    "    # Filepath\n",
    "    #Mouse single-cell RNA-seq data is already published: https://brickman-preimplantation.streamlit.app/Download 3. Raw data - Mouse (01_mouse_reprocessed.h5ad)\n",
    "    file_path = \"/home/.../01_mouse_reprocessed.h5ad\"\n",
    "    # Read h5ad\n",
    "    adata = sc.read_h5ad(file_path)\n",
    "    print(f\"Loaded AnnData object: {adata}\")\n",
    "\n",
    "\n",
    "# Group Stages (Ie. E3.5-ICM and E4.5-ICM in a group ICM. Later we can separate based on \"timepoint\")\n",
    "def map_cell_type(ct):\n",
    "    # Keep early stages without changes\n",
    "    if ct in ['Zygote', '2C', '4C', '8C', '16C']:\n",
    "        return ct\n",
    "    # Combine by Cell type\n",
    "    elif 'TE' in ct:\n",
    "        return 'TE'\n",
    "    elif 'ICM' in ct:\n",
    "        return 'ICM'\n",
    "    elif 'EPI' in ct:\n",
    "        return 'EPI'\n",
    "    elif 'PrE' in ct:\n",
    "        return 'PrE'\n",
    "    else:\n",
    "        return ct  \n",
    "\n",
    "#Save in column ct_merged\n",
    "adata.obs['ct_merged'] = adata.obs['ct'].apply(map_cell_type)\n",
    "\n",
    "# Test for unique values\n",
    "print(adata.obs['ct_merged'].unique())\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def plot_gene_custom_modified(gene, adata, save_path=None, dpi=300):\n",
    "   \n",
    "    # Gene Expression\n",
    "    if gene in adata.var_names:\n",
    "        gene_idx = gene\n",
    "    else:\n",
    "        matches = adata.var.index[adata.var['gene_symbol'].str.lower() == gene.lower()]\n",
    "        if len(matches) > 0:\n",
    "            gene_idx = matches[0]\n",
    "        else:\n",
    "            raise ValueError(f\"El gen {gene} no se encontró en adata.var\")\n",
    "    \n",
    "    expr = adata[:, gene_idx].X\n",
    "    if hasattr(expr, \"toarray\"):\n",
    "        expr = expr.toarray().flatten()\n",
    "    else:\n",
    "        expr = np.array(expr).flatten()\n",
    "    \n",
    "    # Create DataFrame ---\n",
    "    df = pd.DataFrame({\n",
    "        'expression': expr,\n",
    "        'timepoint': adata.obs['timepoint'].astype(str).values,\n",
    "        'ct_merged': adata.obs['ct_merged'].astype(str).values\n",
    "    })\n",
    "    \n",
    "    # Subsets:\n",
    "    early_timepoints = ['Zygote', '2C', '4C', '8C', '16C']\n",
    "    ct_order = [\"ICM\", \"TE\", \"EPI\", \"PrE\"]\n",
    "    #Set Stadistical Comparison Pairs. TE is left out from pairwise comparisons (no special interest in TE, and clearer graph)\n",
    "    pairs = [(\"ICM\", \"EPI\"), (\"ICM\", \"PrE\"), (\"EPI\", \"PrE\")]\n",
    "    \n",
    "    # Subset per each panel:\n",
    "    top_df_early = df[df['timepoint'].isin(early_timepoints)]\n",
    "    df_E3_25 = df[df['timepoint'] == 'E3.25']\n",
    "    df_E3_5  = df[df['timepoint'] == 'E3.5']\n",
    "    df_E4_5  = df[df['timepoint'] == 'E4.5']\n",
    "\n",
    "  \n",
    "    #Set fontsize\n",
    "    plt.rcParams.update({'font.size': 30})\n",
    "    \n",
    "    # Colour palette\n",
    "    # Early Stages\n",
    "    top_palette = {tp: (0.65, 0.65, 0.65) for tp in early_timepoints}\n",
    "    # Other stages: E3.25, E3.5, E4.5\n",
    "     custom_colors = {\n"
    "     \"ICM\": (0.65, 0.65, 0.65),     # grey   \n"
    "     \"TE\": \"#A08376\",              # brown \n"
    "     \"EPI\": \"#693C8C\",             # Magenta\n"
    "     \"PrE\": \"#387138\"              # Dark green\n"
    "     }\n"
    "    \n"
    "    bottom_palette = {ct: custom_colors[ct] for ct in ct_order} \n"
    "    \n",
    "    # Compute means\n",
    "    early_means = top_df_early.groupby(\"timepoint\")[\"expression\"].mean().reset_index()\n",
    "    E3_25_means  = df_E3_25.groupby(\"ct_merged\")[\"expression\"].mean().reset_index()\n",
    "    E3_5_means   = df_E3_5.groupby(\"ct_merged\")[\"expression\"].mean().reset_index()\n",
    "    E4_5_means   = df_E4_5.groupby(\"ct_merged\")[\"expression\"].mean().reset_index()\n",
    "    \n",
    "    # Create 2x2 figure using GridSpec\n",
    "    fig = plt.figure(figsize=(16, 12))\n",
    "    gs = gridspec.GridSpec(nrows=2, ncols=2, height_ratios=[2, 2], hspace=0.4, wspace=0.3)\n",
    "    \n",
    "    bar_width = 0.3  # Ancho de la barra horizontal que marca la media (para paneles manuales)\n",
    "    \n",
    "    # Early Stages\n",
    "    ax1 = fig.add_subplot(gs[0, 0])\n",
    "    sns.violinplot(data=top_df_early, x=\"timepoint\", y=\"expression\", hue=\"timepoint\",\n",
    "                   order=early_timepoints, palette=top_palette, inner=None,\n",
    "                   cut=0, alpha=0.5, dodge=False, legend=False, ax=ax1)\n",
    "    sns.stripplot(data=top_df_early, x=\"timepoint\", y=\"expression\", hue=\"timepoint\",\n",
    "                  order=early_timepoints, palette=top_palette, jitter=True,\n",
    "                  edgecolor=\"black\", linewidth=0.1, dodge=False, legend=False, ax=ax1)\n",
    "    for i, tp in enumerate(early_timepoints):\n",
    "        row = early_means[early_means[\"timepoint\"] == tp]\n",
    "        if not row.empty:\n",
    "            mean_val = row[\"expression\"].values[0]\n",
    "            ax1.hlines(mean_val, i - bar_width/2, i + bar_width/2,\n",
    "                       color=\"black\", linewidth=4, zorder=10)\n",
    "            ax1.text(i, mean_val + 0.2, f\"{mean_val:.1f}\",\n",
    "                     ha=\"center\", va=\"bottom\", fontweight='bold',\n",
    "                     color=\"black\", zorder=10,\n",
    "                     bbox=dict(facecolor='white', alpha=0.2, pad=2, edgecolor='none'))\n",
    "    ax1.set_xticks(range(len(early_timepoints)))\n",
    "    ax1.set_xticklabels(early_timepoints, rotation=45, ha=\"right\")\n",
    "    ax1.set_xlim(-0.5, len(early_timepoints)-0.5)\n",
    "    ax1.set_xlabel(\"\")\n",
    "    ax1.set_ylabel(\"Expression\")\n",
    "    ax1.set_title(\"Early stages\", fontweight='bold')\n",
    "    \n",
    "    # E3.25\n",
    "    ax2 = fig.add_subplot(gs[0, 1])\n",
    "    sns.violinplot(data=df_E3_25, x=\"ct_merged\", y=\"expression\", hue=\"ct_merged\",\n",
    "                   order=ct_order, palette=bottom_palette, inner=None,\n",
    "                   cut=0, alpha=0.5, dodge=False, legend=False, ax=ax2)\n",
    "    sns.stripplot(data=df_E3_25, x=\"ct_merged\", y=\"expression\", hue=\"ct_merged\",\n",
    "                  order=ct_order, palette=bottom_palette, jitter=True,\n",
    "                  edgecolor=\"white\", linewidth=0.1, dodge=False, legend=False, ax=ax2)\n",
    "    # E3.25 order\n",
    "    for j, ct in enumerate(ct_order):\n",
    "        row = E3_25_means[E3_25_means[\"ct_merged\"] == ct]\n",
    "        if not row.empty:\n",
    "            mean_val = row[\"expression\"].values[0]\n",
    "            ax2.hlines(mean_val, j - bar_width/2, j + bar_width/2,\n",
    "                       color=\"black\", linewidth=4, zorder=10)\n",
    "            ax2.text(j, mean_val + 0.2, f\"{mean_val:.1f}\",\n",
    "                     ha=\"center\", va=\"bottom\", fontweight='bold',\n",
    "                     color=\"black\", zorder=10,\n",
    "                     bbox=dict(facecolor='white', alpha=0.2, pad=2, edgecolor='none'))\n",
    "    ax2.set_xticks(range(len(ct_order)))\n",
    "    ax2.set_xticklabels(ct_order, rotation=45, ha=\"right\")\n",
    "    ax2.set_xlim(-0.5, len(ct_order)-0.5)\n",
    "    ax2.set_xlabel(\"\")\n",
    "    ax2.set_ylabel(\"\")\n",
    "    ax2.set_title(\"E3.25\", fontweight='bold')\n",
    "    \n",
    "    # E3.5\n",
    "    ax3 = fig.add_subplot(gs[1, 0])\n",
    "    sns.violinplot(data=df_E3_5, x=\"ct_merged\", y=\"expression\", hue=\"ct_merged\",\n",
    "                   order=ct_order, palette=bottom_palette, inner=None,\n",
    "                   cut=0, alpha=0.5, dodge=False, legend=False, ax=ax3)\n",
    "    sns.stripplot(data=df_E3_5, x=\"ct_merged\", y=\"expression\", hue=\"ct_merged\",\n",
    "                  order=ct_order, palette=bottom_palette, jitter=True,\n",
    "                  edgecolor=\"white\", linewidth=0.1, dodge=False, legend=False, ax=ax3)\n",
    "    for j, ct in enumerate(ct_order):\n",
    "        row = E3_5_means[E3_5_means[\"ct_merged\"] == ct]\n",
    "        if not row.empty:\n",
    "            mean_val = row[\"expression\"].values[0]\n",
    "            ax3.hlines(mean_val, j - bar_width/2, j + bar_width/2,\n",
    "                       color=\"black\", linewidth=4, zorder=10)\n",
    "            ax3.text(j, mean_val + 0.2, f\"{mean_val:.1f}\",\n",
    "                     ha=\"center\", va=\"bottom\", fontweight='bold',\n",
    "                     color=\"black\", zorder=10,bbox=dict(facecolor='white', alpha=0.2, pad=2, edgecolor='none'))       \n",
    "    \n",
    "    # Set Group order in E3.5\n",
    "    full_order = ct_order  # Ejemplo: [\"ICM\", \"TE\", \"EPI\", \"PrE\"]\n",
    "    order_E3_5 = [group for group in full_order if group in df_E3_5['ct_merged'].unique()]\n",
    "    pairs_E3_5 = [pair for pair in pairs if pair[0] in order_E3_5 and pair[1] in order_E3_5]\n",
    "    annotator3 = Annotator(ax3, pairs_E3_5, data=df_E3_5, x=\"ct_merged\", y=\"expression\", order=order_E3_5)\n",
    "    annotator3.configure(test='t-test_ind', text_format='star', hide_non_significant=True, loc='inside',comparisons_correction=\"bonferroni\", verbose=0)\n",
    "    annotator3.apply_and_annotate()\n",
    "    # Get offset: position of the first Group.\n",
    "    if len(order_E3_5) > 0:\n",
    "        offset_E3_5 = full_order.index(order_E3_5[0])\n",
    "    else:\n",
    "        offset_E3_5 = 0\n",
    "    # Adjust positions in ax3:\n",
    "    for txt in [t for t in ax3.texts if t.get_fontweight()=='bold' and t.get_zorder()==10]:\n",
    "        pos = txt.get_position()\n",
    "        txt.set_position((pos[0] + offset_E3_5, pos[1]))\n",
    "\n",
    "    ax3.set_xticks(range(len(full_order)))\n",
    "    ax3.set_xticklabels(full_order, rotation=45, ha=\"right\")\n",
    "    ax3.set_xlim(-0.5, len(full_order)-0.5)\n",
    "    ax3.set_xlabel(\"\")\n",
    "    ax3.set_ylabel(\"Expression\")\n",
    "    ax3.set_title(\"E3.5\", fontweight='bold')\n",
    "\n",
    "\n",
    "    # E4.5\n",
    "    # Filter E4_5 to remove \"ICM\"\n",
    "    df_E4_5 = df[(df['timepoint'] == 'E4.5') & (df['ct_merged'] != \"ICM\")]\n",
    "    # Set order in E4_5 without \"ICM\":\n",
    "    ct_order_E4_5 = [g for g in ct_order if g != \"ICM\"]\n",
    "    \n",
    "    ax4 = fig.add_subplot(gs[1, 1])\n",
    "    sns.violinplot(data=df_E4_5, x=\"ct_merged\", y=\"expression\", hue=\"ct_merged\",\n",
    "                   order=ct_order_E4_5, palette=bottom_palette, inner=None,\n",
    "                   cut=0, alpha=0.5, dodge=False, legend=False, ax=ax4)\n",
    "    sns.stripplot(data=df_E4_5, x=\"ct_merged\", y=\"expression\", hue=\"ct_merged\",\n",
    "                  order=ct_order_E4_5, palette=bottom_palette, jitter=True,\n",
    "                  edgecolor=\"white\", linewidth=0.1, dodge=False, legend=False, ax=ax4)\n",
    "    \n",
    "    for j, ct in enumerate(ct_order_E4_5):\n",
    "        row = E4_5_means[E4_5_means[\"ct_merged\"] == ct]\n",
    "        if not row.empty:\n",
    "            mean_val = row[\"expression\"].values[0]\n",
    "            ax4.hlines(mean_val, j - bar_width/2, j + bar_width/2,\n",
    "                       color=\"black\", linewidth=4, zorder=10)\n",
    "            ax4.text(j, mean_val + 0.2, f\"{mean_val:.1f}\",\n",
    "                     ha=\"center\", va=\"bottom\", fontweight='bold',\n",
    "                     color=\"black\", zorder=10,bbox=dict(facecolor='white', alpha=0.2, pad=2, edgecolor='none'))    \n",
    "\n",
    "    \n",
    "    # Stats in E4_5 without \"ICM\":\n",
    "    pairs_E4_5 = [pair for pair in pairs if pair[0] in ct_order_E4_5 and pair[1] in ct_order_E4_5]\n",
    "    annotator4 = Annotator(ax4, pairs_E4_5, data=df_E4_5, x=\"ct_merged\", y=\"expression\", order=ct_order_E4_5)\n",
    "    annotator4.configure(test='t-test_ind', text_format='star', hide_non_significant=True, loc='inside', verbose=0, comparisons_correction='bonferroni')\n",
    "    annotator4.apply_and_annotate()\n",
    "    \n",
    "    ax4.set_xticks(range(len(ct_order_E4_5)))\n",
    "    ax4.set_xticklabels(ct_order_E4_5, rotation=45, ha=\"right\")\n",
    "    ax4.set_xlim(-0.5, len(ct_order_E4_5)-0.5)\n",
    "    ax4.set_xlabel(\"\")\n",
    "    ax4.set_ylabel(\"\")\n",
    "    ax4.set_title(\"E4.5\", fontweight='bold')\n",
    "\n",
    "\n",
    "    # Make scales even in all 4 graphs\n",
    "    all_axes = [ax1, ax2, ax3, ax4]\n",
    "    max_y = max(ax.get_ylim()[1] for ax in all_axes)\n",
    "    for ax in all_axes:\n",
    "        ax.set_ylim(0, max_y)\n",
    "    \n",
    "    # Global Title\n",
    "    fig.suptitle(f\"{gene}\", fontsize=38, fontweight='bold')\n",
    "    plt.tight_layout(rect=[0, 0, 1, 0.93])\n",
    "\n",
    "    #Save\n",
    "    if save_path is not None:\n",
    "        os.makedirs(os.path.dirname(save_path), exist_ok=True)\n",
    "        plt.savefig(save_path, dpi=dpi)\n",
    "    else:\n",
    "        plt.show()\n",
    "    plt.close(fig)\n",
    "\n",
    "\n",
    "#USAGE:\n",
    "genes_to_plot = [\"nanog\", \"gata6\", \"sparc\", \"gata4\", \"aqp8\", \"dusp4\", \"ctnnb1\"]\n",
    "\n",
    "for gene in genes_to_plot:\n",
    "    out_file = f\"./figures/{gene}.png\"\n",
    "    \n",
    "    plot_gene_custom_modified(gene, adata, save_path=out_file, dpi=300)\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
